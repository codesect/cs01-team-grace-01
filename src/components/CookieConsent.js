import React, { useState } from 'react';
import styled from 'styled-components';

import { Wrapper } from './GlobalStyles';

const Button = styled.button`
  align-items: center;
  background-color: ${({ theme }) => theme.text};
  color: ${({ theme }) => theme.buttonText};
  display: flex;
  justify-content: space-between;
  min-width: 10rem;

  @media (min-width: ${({ theme }) => theme.breakpointS}) {
    margin-left: 1rem;
  }

  &:focus {
    outline-color: black;
  }

  line {
    transform-origin: center;
    transition: transform ${({ theme }) => theme.transitionEase};
  }

  &:focus,
  &:hover {
    line {
      transform: scale(0.5);
    }
  }
`;

const Container = styled.div`
  background-color: ${({ theme }) => theme.consentBackground};
  bottom: 0;
  font-size: 0.75rem;
  font-weight: 300;
  line-height: 1.25;
  position: fixed;
  width: 100%;
`;

const Text = styled.p`
  margin: 0 0 1rem;

  @media (min-width: ${({ theme }) => theme.breakpointS}) {
    margin: 0;
  }
`;

const ConsentWrapper = styled(Wrapper)`
  align-items: center;
  display: flex;
  flex-direction: column;

  @media (min-width: ${({ theme }) => theme.breakpointS}) {
    flex-direction: row;
  }
`;

export function isCookieAccepted() {
  const consent = document.cookie
    .split(';')
    .filter(item => item.includes('cookieConsent=true')).length;

  return Boolean(consent);
}

function CookieConsent() {
  const [isVisible, setIsVisible] = useState(!isCookieAccepted());

  function acceptCookie() {
    setIsVisible(false);

    const inAYear = new Date(
      new Date().getTime() + 365 * 24 * 60 * 60 * 1000,
    ).toUTCString();

    document.cookie = `cookieConsent=true; expires=${inAYear}; path=/`;
  }

  if (!isVisible) {
    return null;
  }

  return (
    <Container>
      <ConsentWrapper>
        <Text>
          By browsing this site you accept cookies used to improve and
          personalise our services. Read our privacy policy for more about what
          we do with your data, and how to manage cookies.
        </Text>
        <Button type="button" onClick={acceptCookie}>
          Accept and Close
          <svg
            width="16"
            height="16"
            viewBox="0 0 360 360"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="95.1665"
              y1="264.853"
              x2="264.872"
              y2="95.1472"
              stroke="currentColor"
              strokeWidth="16"
              strokeLinecap="round"
            />
            <line
              x1="264.872"
              y1="264.853"
              x2="95.1665"
              y2="95.1472"
              stroke="currentColor"
              strokeWidth="16"
              strokeLinecap="round"
            />
            <circle
              cx="180"
              cy="180"
              fill="none"
              r="172"
              stroke="currentColor"
              strokeWidth="16"
            />
          </svg>
        </Button>
      </ConsentWrapper>
    </Container>
  );
}

export default CookieConsent;
