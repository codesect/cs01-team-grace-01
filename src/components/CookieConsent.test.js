import React from 'react';
import { shallow } from 'enzyme';

import CookieConsent, { isCookieAccepted } from './CookieConsent';

describe('isCookieAccepted', () => {
  beforeEach(() => {
    document.cookie = 'cookieConsent=; expires=Thu, 01 Jan 1970 00:00:00 UTC;';
  });

  it('returns true if cookie exists', () => {
    document.cookie = `cookieConsent=true`;

    expect(isCookieAccepted()).toBe(true);
  });

  it('returns false if cookie does not exist', () => {
    expect(isCookieAccepted()).toBe(false);
  });
});

describe('<CookieConsent />', () => {
  beforeEach(() => {
    document.cookie = 'cookieConsent=; expires=Thu, 01 Jan 1970 00:00:00 UTC;';
  });

  it('returns null if user already accepted cookies', () => {
    document.cookie = 'cookieConsent=true';
    const wrapper = shallow(<CookieConsent />);

    expect(wrapper.isEmptyRender()).toBeTruthy();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<CookieConsent />);

    expect(wrapper).toMatchSnapshot();
  });
});
