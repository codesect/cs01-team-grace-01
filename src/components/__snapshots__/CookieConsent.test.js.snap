// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CookieConsent /> matches snapshot 1`] = `
<styled.div>
  <Styled(styled.div)>
    <styled.p>
      By browsing this site you accept cookies used to improve and personalise our services. Read our privacy policy for more about what we do with your data, and how to manage cookies.
    </styled.p>
    <styled.button
      onClick={[Function]}
      type="button"
    >
      Accept and Close
      <svg
        height="16"
        viewBox="0 0 360 360"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          stroke="currentColor"
          strokeLinecap="round"
          strokeWidth="16"
          x1="95.1665"
          x2="264.872"
          y1="264.853"
          y2="95.1472"
        />
        <line
          stroke="currentColor"
          strokeLinecap="round"
          strokeWidth="16"
          x1="264.872"
          x2="95.1665"
          y1="264.853"
          y2="95.1472"
        />
        <circle
          cx="180"
          cy="180"
          fill="none"
          r="172"
          stroke="currentColor"
          strokeWidth="16"
        />
      </svg>
    </styled.button>
  </Styled(styled.div)>
</styled.div>
`;
